docs(ic): added new feature Notifications API as well as guidance and potential benefits of Digital Twins to Industry Core KIT (#1158)
eclipse-tractusx/eclipse-tractusx.github.io: docs-kits/kits/Industry Core Kit/Software Development View/page_notifications.mdx

---
id: Notifications Development View Industry Core Kit
title: Notifications
description: "Industry Core KIT"
sidebar_position: 3
---

<!--
This work is licensed under the CC-BY-4.0 (https://creativecommons.org/licenses/by/4.0/legalcode).

- SPDX-License-Identifier: CC-BY-4.0
- SPDX-FileCopyrightText: 2023 BASF SE
- SPDX-FileCopyrightText: 2023 Bayerische Motoren Werke Aktiengesellschaft (BMW AG)
- SPDX-FileCopyrightText: 2023 Fraunhofer-Gesellschaft zur Foerderung der angewandten Forschung e.V. (represented by Fraunhofer ISST & Fraunhofer IML)
- SPDX-FileCopyrightText: 2023 German Edge Cloud GmbH & Co. KG
- SPDX-FileCopyrightText: 2023 Mercedes Benz AG
- SPDX-FileCopyrightText: 2023 Robert Bosch Manufacturing Solutions GmbH
- SPDX-FileCopyrightText: 2023 SAP SE
- SPDX-FileCopyrightText: 2023 Siemens AG
- SPDX-FileCopyrightText: 2023 T-Systems International GmbH
- SPDX-FileCopyrightText: 2023 ZF Friedrichshafen AG
- SPDX-FileCopyrightText: 2023 Contributors to the Eclipse Foundation
- Source URL: https://github.com/eclipse-tractusx/eclipse-tractusx.github.io/tree/main/docs-kits/kits/Industry%20Core%20Kit (latest version)
-->

> Industry Core covers the foundation for message-based data exchange using notifications. Notifications are already used in several use cases, but their implementation is not yet standardized. Most use cases, although, based their implementation on the first implementation, Traceability Quality Alerts & Investigations (on which Industry Core notifications are also based). Industry Core now defines an interoperable solution for sending and receiving notifications via connectors in Catena-X. 

Reviewer: Please always use "data sharing" instead of "data exchange". Data exchange means: If I give you A, you will give me B. Data sharing is not so specific and also works in one direction.

```suggestion
Industry Core covers the foundation for message-based data sharing using notifications. Notifications are already used in several use cases, but their implementation is not yet standardized. Most use cases, although, based their implementation on the first implementation, Traceability Quality Alerts & Investigations (on which Industry Core notifications are also based). Industry Core now defines an interoperable solution for sending and receiving notifications via connectors in Catena-X. 
```

A general descriptions of notifications is given in the [Architecture View](../page_architecture-view.mdx) of this KIT.

### Process

> A notification is sent from one Catena-X partner - called sender - to another Catena-X partner - called receiver - using simple messages in JSON format via Catena-X connector, as shown in the following figure:

Reviewer 2: ```suggestion
A notification is sent from one Data Space participant - called sender - to another Data Space participant - called receiver - using simple messages in JSON format via the connector, as shown in the following figure:
```

It's only a suggestions, but considering other data spaces initiative which may be interested in this KIT, it might be useful to be a little more generic here.

Author: Well, Catena-X is referenced not only here, but throughout the whole KIT. Just changing it here does not make sense. If we want to do that, this should be a separate PR IMHO.

![Notifications Flow](../assets/notifications_process.svg)

> The exchange is done via Catena-X connectors:

Reviewer 2: ```suggestion
The exchange is done via standardized connectors:
```

  - The receiver provides an asset in the connector to which notifications can be sent.
  - The sender initiates a contract negotiation for this asset to get a contract agreement and dataplane endpoint.
  - After a successful contract negotiation, the sender sends the notification via a HTTP POST request to the receiver via the dataplane endpoint from the contract agreement. 
  - All notification operations are sent via the same asset using the same contract agreement.
  
### Notification Data Format

> This section defines the data format of notifications:

Reviewer: ```suggestion
This section defines the data format for notifications:
```

> - Notifications must be in JSON format. 
> - Notifications must be compliant to the following data format:

Reviewer: ```suggestion
Notifications must be compliant to the following JSON data format:
```

```json
  {
    "header": {
      "messageId": "urn:uuid:f9a97301-a000-44dd-b9d8-78488a40c6bb",
      "context": "IndustryCore-DigitalTwinEventAPI-ConnectToParent:2.0.0",
      "sentDateTime": "2024-07-05T08:13:33.20733Z",
      "senderBpn": "BPNL000000000AAA",
      "receiverBpn": "BPNL000000000ZZZ",
      "expectedResponseBy": "2024-07-08T08:13:33.20733Z",
      "version": "3.0.0"
    },
    "content": {}
  }
```

> - The JSON schema for this schema is available [here](../assets/notification.schema.json).

Reviewer 2: I'm confused here. In my understanding the notification.schema.json is based on  `urn:samm:io.catenax.shared.message_header:3.0.0`. This also seems consistent to the example json. But the schema seem to contain a lot more information?

Author: The notification schema contains the MessageHeaderAspect aspect model, but it adds the content property.  Without this extension, use cases could not add data to the notification. The MessageHeaderAspect aspect model probably needs to be reworked anyway.

- The property `header` must be compliant to the aspect model [MessageHeaderAspect](#31-aspect-model-messageheaderaspect). 
  - The `messageId` of a notification must uniquely identify a single message, therefore it must not be reused. No two notifications must share the same `messageId`. Only if a notification could not be sent because of data transfer errors, it MAY be re-sent with the same `messageId`.
  - It is recommended to use the following format (defined in the MessageHeaderAspect aspect model) for property `context`: `<domain>-<subdomain>-<object>:<[major] version>`
    - `<domain>` should be the name of the use case that defines the notification, e.g., `IndustryCore`.
    - `<subdomain>` should be the name of the notification API, e.g., `DigitalTwinEventAPI`.
    - `<object>` should be the name of the operation for which the notification is used, e.g., `ConnectToParent`.
    - Versioning only refers to major versions in both default and fallback cases. 

>     - This is recommended as it allows the notification receiver to freely choose what technology to use for the backend service. If this information is not encoded into the notifcation's `context` propoerty, a backend service like a message-queue technology will not know which notification operation was invoked by the sender as the endpoint information is not available in this case (compared to a REST API backend service).

Reviewer: In this text, only some of the keys of the JSON data format are explained, but not all of them. Please explain all the keys/values. For example, what is the value inside `version`? The version of the aspect model? The version of the notification?

Author: I don't think that this is the correct way. There is already a aspect model documentation where all properties are explained. I don't want to replicate that. Only differences should be documented here. Otherwise, the aspect model documentation would be redundant.

Reviewer: Ok, thats fine for me.

Identifier of Aspect Model:  `urn:samm:io.catenax.shared.message_header:3.0.0#MessageHeaderAspect` 

Aspect model in GitHub:
- Version 3.0.0: https://github.com/eclipse-tractusx/sldt-semantic-models/tree/main/io.catenax.shared.message_header/3.0.0

#### Example: Aspect Model `MessageHeaderAspect 3.0.0`

```json
{
  "header": {
    "messageId": "urn:uuid:f9a97301-a000-44dd-b9d8-78488a40c6bb",
    "context": "IndustryCore-DigitalTwinEventAPI-ConnectToParent:3.0.0",
    "sentDateTime": "2024-07-05T08:13:33.20733Z",
    "senderBpn": "BPNL000000000AAA",
    "receiverBpn": "BPNL000000000ZZZ",
    "expectedResponseBy": "2024-07-08T08:13:33.20733Z",
    "version": "3.0.0"
  },
}
```

### Notification API

The notification API of a use case must be defined with an OpenAPI specification. In this specification, all artefacts of the API are defined, e.g., name of the API, operation names, and notification data formats.

#### Preconditions and Dependencies

> - The use case must standardize the notification API's name. The name of the notification API (used in property `http://purl.org/dc/terms/type`) must be defined in the Catena-X taxonomy published under https://w3id.org/catenax/taxonomy.

Reviewer: If I want to add a new name to the taxonomy, how can I do it? I would suggest to add a hint about this, here.

Author: Create a PR for that file. I would rather link to a documentation for that taxonomy that explain it myself, but there is none as far as I know. But I am not the owner of this, so I would not like to add that here.

- For a notification API, only the HTTP method POST is allowed.

#### API Specification

- Notification operations are modelled as endpoints (resources). For every operation, exactly one endpoint must be defined. This endpoint only supports the HTTP POST method.
- The path URL should be the operation name with optional "-" as word separator. 
- The operationId must be the notification operation name. The same name should be used as `<object>` in the `context` property of the notification header. This redundancy is recommended to allow the notification receiver to freely choose what technology to use for the backend service, e.g, a REST API or message-queue technology.
- The notification is the actual request body of the HTTP POST method. Its schema must also be defined in the specification.
  

> **Example:** [OpenAPI specification for Digital Twin Event API](https://eclipse-tractusx.github.io/api-hub/eclipse-tractusx.github.io/3.0.0/example-digital-twin-event-api.yaml) 

Reviewer: Can you please change the link to a relative link instead of the `https://eclipse-tractusx.github.io` domain?

| 200  | OK | Indicates that the request has succeeded. |
| 4xx  | ERROR | The notification cannot be processed. |
| 5xx  | ERROR | The notification cannot be processed. |

A response body MAY also be defined (e.g., see Feedback notification operation in the example above).

These error codes are currently optional as the Catena-X connector does not yet return correct error messages from backend services. These error codes will become mandatory once the connector returns correct error messages from backend services.

#### Connector Asset Structure

For a notification API, exactly one asset in the connector must be defined. All notifications operations of the API must be invoked through this asset. 

- The property `http://purl.org/dc/terms/type` must reference the name of the notification API as defined in the Catena-X taxonomy published under https://w3id.org/catenax/taxonomy.

**Example of an asset definition for a notification API:**

```json
{
  "@id": "DigitalTwinEventAPI",
  "@type": "Asset",
  "properties": {
    "dct:type": {

>       "@id": "cx-taxo:DigitalTwinEventAPI"

Reviewer 2: `cx-taxo:DigitalTwinEventAPI`
does not seem to exist in the taxonomy?

  "@context": {
    "dct": "http://purl.org/dc/terms/",
    "cx-taxo": "https://w3id.org/catenax/taxonomy#",
    "cx-common": "https://w3id.org/catenax/ontology/common#"
  }
}
```

#### Versioning

Versioning of the notification API must follow the Catena-X guidelines for versioning. For every new major version of a notification API, a new connector asset must be created. If, e.g., a REST API is used as backend notification receiver, the incoming notifications can be fowarded to the correct REST API version via the corresponding connector asset, as depicted in the following figure:

![Notifications Flow](../assets/notifications_connector_versioning.svg)

The API's version is not only documented in the notification's connector asset (property `https://w3id.org/catenax/ontology/common#version`), but also in the OpenAPI specification as well as in a notification's `context` property.

### Notification API Documentation

A use case defining a notification API as part of its standard must document the following artifacts:
- All notification data formats must be documented either as separate JSON schema or (also in JSON schema) as part of the notification API's OpenAPI specification.

> - The notification API must be documented with an OpenAPI specification that must be published on the Tractus-X Swagger space.

Reviewer: Where can I find the "Tractus-X Swagger space"? Please link it here.

Author: Based on this documentation: https://eclipse-tractusx.github.io/docs/release/trg-1/trg-1-08/

Author: I changed the wording to "and must be published according to [TRG 1.08 - Interface documentation (APIs)](/docs/release/trg-1/trg-1-08/)"

> - For a standardized use case, the name of the notification API (used in property `http://purl.org/dc/terms/type`) must be defined in the Catena-X taxonomy published under https://w3id.org/catenax/taxonomy.

Reviewer: This is a duplicate sentence and is also mentioned above. I would suggest removing one of them.
